File: Python (Assignment 1.2 - Object-oriented)/PPMImage.py
from BaseImage import BaseImage


class PPMImage(BaseImage):
    def __init__(self, dimensions, themes, binary_format='P3', max_color_val=255):
        super().__init__(dimensions, themes, binary_format)
        self._maxColorVal = max_color_val

    def read(self, file_name):
        with open(file_name, 'r') as file:
            for line in file:
                print(line)

    def draw(self, file_name):
        self._themes.init(self, self._patterns)
        self._choose_pattern().create()
        self._write_to(file_name)

    def _choose_pattern(self):
        print('Choose # from the list of built-in patterns')
        for pattern in self._patterns:
            print(pattern)
        choice = int(input('Input: '))
        if choice <= -1:
            print('CHOICES START AT 0. By default, pattern 0 is chosen for you.')
            return self._patterns[0]

        return self._patterns[choice]

    def _write_to(self, file_name):
        print('Writing to file...')

        with open(file_name, 'w') as file:
            file.write(f'{self._binaryFormat}\n'
                       f'{self._Dimensions.width} {self._Dimensions.height}\n'
                       f'{self._maxColorVal}\n')

            for pixel in self._pixels:
                file.write(str(pixel))

        print("PPM Created")

File: C++ (Assignment 1.2 - Object-oriented)/PPMImage.cpp
#include <fstream>
#include <string>
#include <iostream>

#include <thread>
#include <chrono>

#include "PPMImage.h"

PPMImage::PPMImage(dimensions Dimensions, patternFactory* themes, std::string binaryFormat, int maxColorVal) 
	: baseImage(Dimensions, themes, binaryFormat), maxColorVal(maxColorVal){}

void PPMImage::writeTo(std::string fileName) {
	std::fstream file;	
	file.open(fileName, std::ios::out);

	if (file.is_open()) {
		std::cout << "Writing to file...\n";

		file << this->binaryFormat << "\n"
			<< this->Dimensions.width << " " << this->Dimensions.height << "\n"
			<< this->maxColorVal << "\n";

		for (auto itr{ this->pixels.begin() }; itr != this->pixels.end(); ++itr) {
			itr != (this->pixels.end() - 1) ? file << *(*(itr)) << "\n" : file << *(*itr);
		}
	}
	system("cls");
	std::cout << "PPM Created\n";
}

void PPMImage::read(std::string fileName)  {
	std::fstream file;
	file.open(fileName, std::ios::in);

	if (file.is_open()) {

		std::string text;
		while (std::getline(file, text)) {
			std::cout << text << "\n";
		}
	}
	file.close();
}

void PPMImage::draw(std::string fileName) {
	themes->init(this, &patterns);
	choosePattern().create();
	this->writeTo(fileName);
}

pattern& PPMImage::choosePattern() {
	
	int choice{ 0 };
	try {
		std::cout << "Choose # from the list of built-in patterns:\n";
		for (auto& pattern : patterns) std::cout << *pattern << std::endl;
		std::cout << "\nInput: ";
		std::cin >> choice;
		system("cls");
		if (choice <= -1) throw(choice);
	} 
	catch (int invalidChoice) {
		std::cout << "CHOICES START AT 0. By default, pattern 0 is chosen for you.\n";
		std::this_thread::sleep_for(std::chrono::seconds(3));
		system("cls");
		return *patterns[0];	
	}

	return *patterns[choice];
}